geomesa-kafka/geomesa-kafka-confluent/src/main/scala/org/locationtech/geomesa/kafka/confluent/ConfluentFeatureSerializer.scala

```diff
@@ -49,60 +43,42 @@ class ConfluentFeatureSerializer(
     val options: Set[SerializationOption] = Set.empty
 ) extends SimpleFeatureSerializer with LazyLogging {
 
-  private val visAttributeIndex = sft.indexOf(ConfluentFeatureSerializer.VisAttributeName)
-
   private val kafkaAvroDeserializer = new ThreadLocal[KafkaAvroDeserializer]() {
     override def initialValue(): KafkaAvroDeserializer = new KafkaAvroDeserializer(schemaRegistryClient)
   }
 
-  private var geomSrcAttributeName: Option[String] = None
-
-  def deserialize(id: String, bytes: Array[Byte], date: Date): SimpleFeature = {
-    val genericRecord = kafkaAvroDeserializer.get.deserialize("", bytes).asInstanceOf[GenericRecord]
-    val attrs = sft.getAttributeDescriptors.asScala.map(_.getLocalName).map {
-      case ConfluentFeatureSerializer.GeomAttributeName =>
-        geomSrcAttributeName match {
-          case None =>
-            // Here we find a valid geom field in the first record or throw.
-            val kv = sft.getAttributeDescriptors.asScala
-                .iterator
-                .map(_.getLocalName)
-                .map(n => n -> readFieldAsWkt(genericRecord, n, logFailure = false))
-                .collectFirst { case (k, Some(v)) => k -> v }
-                .getOrElse {
-                  throw new UnsupportedOperationException("No valid WKT field found in avro data for " +
-                      s"ConfluentFeatureSerializer in first record $genericRecord.  Valid Geometry field is required.")
-                }
-            geomSrcAttributeName = Option(kv._1)
-            kv._2
-
-          case Some(name) => readFieldAsWkt(genericRecord, name, logFailure = true).get
-        }
+  override def deserialize(id: String, bytes: Array[Byte]): SimpleFeature = {
+    val record = kafkaAvroDeserializer.get.deserialize("", bytes).asInstanceOf[GenericRecord]
 
-      case ConfluentFeatureSerializer.DateAttributeName => date
-      case name => genericRecord.get(name)
-    }
-    val sf = ScalaSimpleFeature.create(sft, id, attrs: _*)
-    if (visAttributeIndex != -1) {
-      SecurityUtils.setFeatureVisibility(sf, sf.getAttribute(visAttributeIndex).asInstanceOf[String])
-    }
-    sf
-  }
+    var visibilityAttributeName: Option[String] = None
 
-  override def deserialize(id: String, bytes: Array[Byte]): SimpleFeature = deserialize(id, bytes, null)
+    val attributes = sft.getAttributeDescriptors.asScala.map { descriptor =>
```

Comment:
sure, basically when the serializer is created you already have the simple feature type, so you already know the paths each of the `if` checks will take. So you would go through all the descriptors up front and you would create a reader instance that just does the exact call you need for each attribute.

There are examples of something similar in https://github.com/locationtech/geomesa/blob/main/geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/AvroFeatureSerializer.scala#L65 and https://github.com/locationtech/geomesa/blob/main/geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/FeatureSpecificReader.scala

Reply:
So to do this would I need to implement my own `Decoder` to pass into `FeatureSpecificReader.read()`?
And I would store the reader along with the SFT for each topic?

Reply:
i think the decoders are just part of avro. you'd basically want to create 'readers' for each attribute similarly to this function: https://github.com/locationtech/geomesa/blob/main/geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/FeatureSpecificReader.scala#L143
then when you deserialize, you just run through them all like: https://github.com/locationtech/geomesa/blob/main/geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/FeatureSpecificReader.scala#L136

so you'd have a 'reader' for WKT geoms, a different 'reader' for WKB geoms, a 'reader' for millisecond dates, a 'reader' for string formatted dates, etc. when you initialize this serializer class for a given sft, you calculate all the readers to use for that particular sft up front.

Reply:
This kind of what you had in mind? https://github.com/locationtech/geomesa/pull/2823/files#diff-5afc8a29a1528da8edb3edbfa26d7c8cded3fccc8f575e3f5ed340cec7be8351R104